Added design changes for elective labels

arslanashraf7 opened the pull request:
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Mobile width screenshots

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Fixes #4660 

#### What's this PR do?
This PR updated the design for elective courses on the program details page. Added required elective courses count details.

#### How should this be manually tested?
Go to micromasters home page and select a program. See if the design matches the one provided in the associated ticket. 

#### Screenshots (if appropriate)

**New desktop screenshot:**
<img width="1437" alt="Screenshot 2020-10-28 at 4 06 55 PM" src="https://user-images.githubusercontent.com/34372316/97428460-d4432d00-1937-11eb-94b6-245f9a32b9f7.png">

**Old desktop screenshot:**
<img width="1437" alt="Screenshot 2020-10-26 at 5 49 53 PM" src="https://user-images.githubusercontent.com/34372316/97174281-b0ef7500-17b3-11eb-9548-a34a5a71fbdc.png">

**New mobile screenshot:**
<img width="423" alt="Screenshot 2020-10-28 at 4 07 47 PM" src="https://user-images.githubusercontent.com/34372316/97428484-d9a07780-1937-11eb-92f6-7413d558ea3a.png">

**Old mobile screenshot:**
<img width="413" alt="Screenshot 2020-10-26 at 5 51 45 PM" src="https://user-images.githubusercontent.com/34372316/97174459-f3b14d00-17b3-11eb-8f6e-0cb8bb8a40b2.png">

pdpinch commented on cms/serializers_test.py:
```diff
@@ -58,6 +59,8 @@ def test_program_page_serializer(self):
         page = ProgramPageFactory.create()
         courses = CourseFactory.create_batch(3, program=page.program)
         faculty = FacultyFactory.create_batch(3, program_page=page)
+        elective_set = ElectivesSet.objects.create(program=page.program, required_number=1)
```
For what it's worth, we have two known scenarios: "Complete 1 of 2" and "Complete 2 of 3"

arslanashraf7 replied:
Sure, I'll update the test with the above-mentioned scenario, Or should I go with creating a new test? What would you say?

pdpinch replied:
I think you should update it. I don't see any point in testing a "Complete 1 of 1" use case, because that's never going to happen.

arslanashraf7 commented on cms/serializers.py:
```diff
@@ -48,6 +49,13 @@ class ProgramPageSerializer(serializers.ModelSerializer):
+        """Get number of required elective courses for a program"""
+        if programpage.program and programpage.program.electives_set.exists():
+            return programpage.program.electives_set.all().aggregate(Sum('required_number'))['required_number__sum']
+        return 0
```
@annagav  Actually, I did test it again with multiple elective sets and it works for me. Since programs are associated with elective sets using a foreign key, the part where ```program.electives_set``` is used iterates with a reverse foreign key and it contains all the elective sets associated with a program.

**Scenario I tested:**
1. Created an elective set with required number value 1.
2. Created another elective set with required number value 1.
3. Associated both of the sets with a single/same program.

The above query returns the right value summing up the required number field from all the associated sets. e.g. 2 in the above scenario. I have also tested with other scenarios changing the required number value and it seems to work.

Can you please give it a try as well, perhaps this might work for you as well.

annagav replied:
I don't think that we want to display the sum of all required courses over all sets. 
If there are multiple sets, for each set we should be showing it's required number of courses.
So if you have two elective sets, in the program page it should right:
electives (1 of 2):
course 1
course 2
electives (2 of 3):
course 3
course 4 
course 5

So in other words the two sets should be kept separate from each other.

arslanashraf7 replied:
This might be a little different in our case I suppose. Previously, The implementations of the course list used to mark all the elective courses from all the associated elective sets cumulatively. And as per the new design link attached in the associated issue, there is no set wise grouping but course type wise (Core or Elective). 

Maybe we can refer to @pdpinch for this change, we might need the updated design if we are going to group the courses by the elective set as well.

pdpinch replied:
Sorry, I don't think I understand. Anna's comment makes sense to me, but I don't know what might have been done previously. 

To go back to what is needed. We have 
1. the DEDP program, which has 3 core courses and then the learner much complete 2 of 3 electives. You can see this on prod now at https://micromasters.mit.edu/dedp/
2. a new program, which will have 4 core courses and then learner must complete 1 of 2 electives. 

We don't have any plans at the moment for having more than 2 sets for a course (1 core and 1 elective), but I suppose it could happen in the future.

arslanashraf7 replied:
Thanks for the details. I'll go ahead and add these changes.

annagav reviewed (changes requested):
The code looks good! The only thing about the design, I can't see the updated designs, but I would prefer keeping the word ELECTIVE all capitalized, and the title of the set is not needed, I added it to the model mostly for "just in case". I think it is unlikely that the set could be named in any meaningful way, that would be worth showing on the program page.

arslanashraf7 commented:
> The code looks good! The only thing about the design, I can't see the updated designs, but I would prefer keeping the word ELECTIVE all capitalized, and the title of the set is not needed, I added it to the model mostly for "just in case". I think it is unlikely that the set could be named in any meaningful way, that would be worth showing on the program page.

@annagav I agree that the design link on the associated issue isn't updated with multiple Electives Sets, but just as a side note, I had a conversation with @pdpinch on how we are going to show the electives headers and we decided to use the title field there.
I also added these additional details putting a comment on the issue which could be found https://github.com/mitodl/micromasters/issues/4660#issuecomment-717542291.
Additionally, I think with the current implementation we have the flexibility and full control of naming the title as we like and maybe we can use this implementation in our favor rather than hardcoding `ELECTIVE` as a header. Even if we want to show the title just as `ELECTIVE`. We will just need to give `ELECTIVE` in the Elective Set's title field and with the current implementation, we can change this in the future without changing the code if we want to give some other name to the header.

Let me know your thoughts.

pdpinch reviewed (approved):
👍
